### PR TITLE
DEV: add db:validate_indexes task

### PR DESCRIPTION
**Pull request by SamSaffron**
Added a new task to test if indexes are coherent with a blank database

This allows us to detect for cases where somehow indexes are out of sync

### CODE REVIEW

File: lib/tasks/db.rake [hunk 1]
```diff
@@ -295,6 +295,292 @@ task 'db:stats' => 'environment' do
...
+        @pg_bin_path = bin_path
+        break
+      end
+    end
+    if !@pg_bin_path
+      bin_path = "/Applications/Postgres.app/Contents/Versions/latest/bin"
+      if File.exists?("#{bin_path}/pg_ctl")
+        @pg_bin_path = bin_path
+      end
+    end
```
Comment (tgxworld) [inline lib/tasks/db.rake:325]: This is probably minor but I wonder if we can use `RUBY_PLATFORM` is detect which platform the rake task is being executed on to figure out where the bin path is.

File: lib/tasks/db.rake [hunk 2]
```diff
@@ -295,6 +295,292 @@ task 'db:stats' => 'environment' do
...
+    @pg_ctl_path = `which pg_ctl 2> /dev/null`.strip
+    if @pg_ctl_path.length == 0
+      @pg_ctl_path = "#{pg_bin_path}/pg_ctl"
+    end
+
+    @pg_ctl_path
+  end
+
+  def start
+    FileUtils.rm_rf PG_PATH
```
Comment (tgxworld) [inline lib/tasks/db.rake:366]: Minor would name this something like `PG_TMP_PATH`. On the first look I thought we were moving a folder within the PG data directory.

File: lib/tasks/db.rake [hunk 3]
```diff
@@ -295,6 +295,292 @@ task 'db:stats' => 'environment' do
...
+
+    puts "Starting postgres on port: #{pg_port}"
+    ENV['DISCOURSE_PG_PORT'] = pg_port.to_s
+
+    Thread.new do
+      `#{pg_ctl_path} -D '#{PG_PATH}' start`
+    end
+
+    puts "Waiting for PG server to start..."
+    while !`#{pg_ctl_path} -D '#{PG_PATH}' status`.include?('server is running')
```
Comment (tgxworld) [inline lib/tasks/db.rake:381]: We can use [pg_isready](https://www.postgresql.org/docs/13/app-pg-isready.html) too

File: lib/tasks/db.rake [hunk 4]
```diff
@@ -295,6 +295,292 @@ task 'db:stats' => 'environment' do
...
+              begin
+                DB.exec("DROP INDEX #{index_name}")
+              rescue => e
+                $stderr.puts "Error dropping index #{index_name} - #{e}"
+              end
+            else
+              $stderr.puts "Skipping #{index_name} since #{table_name} should not exist - maybe an old plugin created it"
+            end
+          else
+            $stderr.puts "ERROR - BAD REGEX - UNABLE TO PARSE INDEX"
```
Comment (tgxworld) [inline lib/tasks/db.rake:570]: Probably useful to know which index the regexp can't parse.

File: lib/tasks/db.rake (hunk 1 above)
Comment (SamSaffron) [reply lib/tasks/db.rake:325]: Yeah I am not sure I even want to support this outside of our docker install. Considering removing all the mac stuff here.

File: lib/tasks/db.rake (hunk 2 above)
Comment (SamSaffron) [reply lib/tasks/db.rake:366]: sure renaming

Comment (SamSaffron) [reply lib/tasks/db.rake:366]: sure done.

File: lib/tasks/db.rake (hunk 3 above)
Comment (SamSaffron) [reply lib/tasks/db.rake:381]: hmmm agree, let me try to get it working.

File: lib/tasks/db.rake (hunk 2 above)
Comment (SamSaffron) [reply lib/tasks/db.rake:366]: actually it kind of complicates the code a tiny bit, going to keep ti using pg ctl for now. 

worse case it breaks one day and just hangs there. will not be a giant problem.

**Review by tgxworld — APPROVED** [review]

**Pull request merged**